### DISCUSSION TITLE
Displaying output from script in a RichLog line by line

### DISCUSSION THREAD

**Discussion by Darkstarinternet** [Textualize/textual#3788] [Q&A] [answered]
I have a very simple zsh script called subprocess_script which looks like this:

```
#!/bin/zsh

for i in {0..3}
do
  echo "Number: $i"
  sleep 1
done
```

I am trying to run the script as a subprocess and displaying the output in a Textual RichLog widget. 

My Textual code is:

```
import subprocess

from textual import on
from textual.app import App, ComposeResult
from textual.widgets import Footer, RichLog, Button


class SubprocessMessagesApp(App[None]):
    BINDINGS = [
        ("q", "quit", "Quit"),
    ]

    def __init__(self) -> None:
        super().__init__()
        self.results_log = RichLog()
        self.run_button = Button("Run")

    def compose(self) -> ComposeResult:
        yield self.run_button
        yield self.results_log
        yield Footer()

    @on(Button.Pressed)
    def on_button_pressed(self) -> None:
        self.run_subprocess("subprocess_script")

    def run_subprocess(self, script_path):
        with subprocess.Popen(
            ["/bin/zsh", script_path], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True
        ) as proc:
            if proc.stdout is not None:
                for line in proc.stdout:
                    self.results_log.write(line.strip())


if __name__ == "__main__":
    subprocess_messages_app = SubprocessMessagesApp()
    subprocess_messages_app.run()
```

What actually happens when you press "Run" is that there is a three second pause before all the output is displayed at once. What I want to happen is for each line to be output one by one with a one second delay between each line being displayed in the RichLog. I can't figure out how to do this when using Textual, I have it working from a non textual python file. Thanks for any assistance.

**Comment by TomJGooding** [comment, marked as answer]
> If you're going to be doing non-async-friendly busy work, and you want to update the UI as the work takes place, you should have a read about [workers](https://textual.textualize.io/guide/workers/) and, in your case, especially [thread workers](https://textual.textualize.io/guide/workers/#thread-workers).

https://github.com/Textualize/textual/discussions/3776#discussioncomment-7700713
